### PR TITLE
MWPW-186777: Add per country promo codes and offer selection for promotions

### PR DESCRIPTION
**Pull request by Andrei4226**
Resolves https://jira.corp.adobe.com/browse/MWPW-186777
QA Checklist: https://wiki.corp.adobe.com/display/adobedotcom/M@S+Engineering+QA+Use+Cases

- Added per country promo code overrides and offer substitutions management (`mas-promo-codes-manager` dialog)
- Added `Offers` tab to the promotions item selector/table, offers selected via OST, metadata cached in the offers field
- Offer product tags auto-filter fragment search when offers are selected
- Removing an offer prompts confirmation and prunes orphaned fragments
- Added promotion duplicate action with title confirmation dialog
- Replaced ad-hoc publish button logic with `canPublishPromotionNow / canSchedulePromotion` helpers
- Added empty state for items section before any offer/fragment is selected

URL for test: https://mwpw-186777--mas--adobecom.aem.live/studio.html#page=promotions-editor&path=sandbox&promotionId=b485c419-f3a1-4226-a950-a8ef0c00197b

Regression

Preview

- non regression on us plans page https://main--cc--adobecom.aem.page/creativecloud/plans?maslibs=MWPW-186777
- non regression on fr plans page https://main--cc--adobecom.aem.page/fr/creativecloud/plans?maslibs=MWPW-186777
- non regression on catalog page https://main--cc--adobecom.aem.page/products/catalog?maslibs=MWPW-186777

mas/io

- non regression on us plans page https://main--cc--adobecom.aem.live/creativecloud/plans?mas-io-url=https://14257-merchatscale-andrei4226.adobeioruntime.net/api/v1/web/MerchAtScale
- non regression on fr plans page https://main--cc--adobecom.aem.live/fr/creativecloud/plans?mas-io-url=https://14257-merchatscale-andrei4226.adobeioruntime.net/api/v1/web/MerchAtScale
- non regression on catalog page https://main--cc--adobecom.aem.live/products/catalog?mas-io-url=https://14257-merchatscale-andrei4226.adobeioruntime.net/api/v1/web/MerchAtScale

Please do the steps below before submitting your PR for a code review or QA

- [x] C1. Cover code with Unit Tests
- [ ] C2. Add a Nala test (double check with #fishbags if nala test is needed)
- [x] C3. Verify all Checks are green (unit tests, nala tests)
- [x] C4. PR description contains working Test Page link where the feature can be tested
- [x] C5: you are ready to do a demo from Test Page in PR (bonus: write a working demo script that you'll use on Thursday, you can eventually put in your PR)
- [x] C.6 read your Jira one more time to validate that you've addressed all AC's and nothing is missing

## Test URLs:

- Before: https://main--mas--adobecom.aem.live/
- After: https://mwpw-186777--mas--adobecom.aem.live/

[wiki](https://wiki.corp.adobe.com/pages/viewpage.action?pageId=3865554764&spaceKey=WP4&title=M%2BS%2BPricing%2BSource%2Bof%2BTruth)

### Regional Pricing
https://mwpw-186777--mas--adobecom.aem.live/studio.html#page=promotions-editor&path=nala&promotionId=68d87504-4ac3-411a-a4e5-97088c2d7494&query=4a466a3c-efa2-4406-ae47-93abd2167e27

page: https://mwpw-186777--mas--adobecom.aem.live/web-components/docs/plans-collection.html?locale=es_CO&mas.preview=on&maslibs=mwpw-186777&debug.io&clearCaches.io

### Intro Pricing
https://mwpw-186777--mas--adobecom.aem.live/studio.html#page=promotions-editor&path=nala&promotionId=11f2b929-9397-4b6d-8055-d76fe05e2d83

page: https://mwpw-186777--mas--adobecom.aem.live/web-components/docs/plans-collection.html?locale=en_AU&maslibs=mwpw-186777&mas.preview=on&debug.io&clearCaches.io

### CODE REVIEW

File: io/www/src/fragment/transformers/customize.js [hunk 1]
```diff
@@ -206,6 +206,21 @@ function mergeVariations(root, customizeContext) {
...
+        fragment.fields.osi = fragOsi.map((o) => {
+            const sub = substituteMap[o];
+            if (sub) logDebug(() => `Substituting OSI ${o} with ${sub} on fragment ${fragment.id}`, context);
+            return sub ?? o;
+        });
+    } else if (substituteMap[fragOsi]) {
+        logDebug(() => `Substituting OSI ${fragOsi} with ${substituteMap[fragOsi]} on fragment ${fragment.id}`, context);
+        fragment.fields.osi = substituteMap[fragOsi];
+    }
+}
```
Comment (npeltier) [inline io/www/src/fragment/transformers/customize.js]: this won't be enough! there are osi used in some content, like text or something. I'd suggest "marking" that fragment with the new osi, and in later transformer WCS, when osi is found, before fetch is made, make the substitution in the content, and let the new one being fetched.

File: studio/src/promotions/promotion-editor-utils.js [hunk 2]
```diff
@@ -162,3 +178,825 @@ export function getPromotionRequiredFieldsValidation(fragment, itemCount) {
...
+    selectedCards,
+    selectedCollections,
+    offerDataCache,
+    cardsByPaths,
+    collectionsByPaths,
+    groupedVariationsByParent,
+    groupedVariationsData,
+}) {
+    const currentCards = selectedCards || [];
+    const currentCollections = selectedCollections || [];
```
Comment (npeltier) [inline studio/src/promotions/promotion-editor-utils.js:259]: is it me or it's not used (yet)?

File: studio/src/promotions/promotion-editor-utils.js [hunk 3]
```diff
@@ -162,3 +178,825 @@ export function getPromotionRequiredFieldsValidation(fragment, itemCount) {
...
+}
+
+export function pruneOrphanedPromotionSelectionAfterOfferRemoval({
+    selectedCards,
+    selectedCollections,
+    remainingSelectedOfferIds,
+    offerDataCache,
+    cardsByPaths,
+    collectionsByPaths,
+    groupedVariationsByParent,
```
Comment (npeltier) [inline studio/src/promotions/promotion-editor-utils.js:308]: dep of my comment above

File: studio/src/promotions/promotion-editor-utils.js [hunk 4]
```diff
@@ -162,3 +178,825 @@ export function getPromotionRequiredFieldsValidation(fragment, itemCount) {
...
+    return String(value).toUpperCase();
+}
+
+/**
+ * Builds mas tag entries for promotion offer table columns from an OST offer payload.
+ * @param {object} offer
+ * @param {string} [productArrangementCode]
+ * @returns {Array<{ id: string, title: string }>}
+ */
+export function buildPromotionOfferTags(offer, productArrangementCode) {
```
Comment (npeltier) [inline studio/src/promotions/promotion-editor-utils.js]: don't we have similar process already existing in card editor? if yes, let's factorize it

File: studio/src/promotions/promotion-editor-utils.js [hunk 5]
```diff
@@ -162,3 +178,825 @@ export function getPromotionRequiredFieldsValidation(fragment, itemCount) {
...
+        .toLowerCase()
+        .replace(/[^a-z0-9]/g, '');
+    if (productName) {
+        const byProductName = ADOBE_PRODUCTS.find(
+            (product) => product.name.toLowerCase().replace(/[^a-z0-9]/g, '') === productName,
+        );
+        if (byProductName) return `${PRODUCT_ICON_BASE_URL}/${byProductName.id}.svg`;
+    }
+
+    return undefined;
```
Comment (npeltier) [inline studio/src/promotions/promotion-editor-utils.js]: same here

File: studio/src/promotions/promotion-editor-utils.js [hunk 6]
```diff
@@ -162,3 +178,825 @@ export function getPromotionRequiredFieldsValidation(fragment, itemCount) {
...
+    const persistedCache = parsePromotionOfferCacheLines(offerValues);
+    const cacheLines = (selectedOfferIds || [])
+        .filter(Boolean)
+        .map((id) => {
+            const snapshot = serializePromotionOfferCacheSnapshot(offerDataCache?.get?.(id)) ?? persistedCache.get(id) ?? null;
+            return snapshot ? serializePromotionOfferCacheLine(id, snapshot) : '';
+        })
+        .filter(Boolean);
+    return serializePromotionOffersField(promoExceptions, offerSubstitutions, selectedOfferIds, cacheLines);
+}
```
Comment (npeltier) [inline studio/src/promotions/promotion-editor-utils.js]: nit: i understand you are building an object out of the serialization of it. However could i ask you _not_ to use words snapshot & cache, that are largely used in other places with a very different meaning. This will add to an already confusing situation

File: studio/src/promotions/mas-promotions-editor.js [hunk 7]
```diff
@@ -7,37 +7,88 @@ import StoreController from '../reactivity/store-controller.js';
...
+    isPromotionOffersSelectionDirty,
     getPromotionRequiredFieldsValidation,
     parsePromotionSurfacesFieldValues,
+    parseCountriesFromGeos,
+    parsePromoCodeExceptions,
+    parsePromotionOffersField,
+    parseSelectedOfferIdsFromOffersField,
+    groupCountriesByPromoCode,
+    handlePromotionOstOfferSelect,
     serializePromotionSurfacesForAem,
```
Comment (npeltier) [inline studio/src/promotions/mas-promotions-editor.js:34]: if promotion-editor-utils function are just meant to be used in promotion editor, i would propose we move them in?

File: studio/src/promotions/mas-promotions.js [hunk 8]
```diff
@@ -3,16 +3,21 @@ import { repeat } from 'lit/directives/repeat.js';
...
+import './mas-promotion-duplicate-dialog.js';
 import { renderPromotionStatusCell } from '../common/utils/render-utils.js';
 import {
+    canPublishPromotionNow,
+    canSchedulePromotion,
     confirmPublishDespiteUnpublishedPromoVariations,
     isPromotionExpiredForPublish,
     publishPromotionProject,
     PROMOTION_EXPIRED_PUBLISH_MESSAGE,
 } from './promotion-publish-utils.js';
```
Comment (npeltier) [inline studio/src/promotions/mas-promotions.js:19]: same here, i wonder i we shouldn't move all promotion publication capabilities to mas-promotions directly

**Review by npeltier — CHANGES_REQUESTED** [review]
2 main requested changes, rest is more "cosmetic":
- osi substitutions should be made also within the content of the fragment. Fragment's osi is here for "categorizing" it. It's later used in card edition to set prices & all. Those prices need to be updated.
- we'd need to double check if it's MVP, but once i added an offer to my project, i should be able to add another one. 
image
i understand that this kind of make the fragment search complicated (maybe a dropdown applying a tag set or another depending on which offer are you search fragments for)

but yeah, you could have e.g. promo project for CCI & Photoshop.

File: io/www/src/fragment/transformers/customize.js (hunk 1 above)
Comment (Andrei4226) [reply io/www/src/fragment/transformers/customize.js]: Done.
`applyOsiSubstitution` was removed from `customize.js`. 
Substitution is now applied in `wcs.js` via `substituteOsi()`.
`customize.js` only uses the effective substituted OSI when looking up promo codes in `applyPromoCode`.

File: studio/src/promotions/promotion-editor-utils.js (hunk 2 above)
Comment (Andrei4226) [reply studio/src/promotions/promotion-editor-utils.js:259]: It is used, imported and called in `mas-promotions-items-table.js` (lines 855 and 892 for `pruneOrphanedPromotionSelectionAfterOfferRemoval`) as part of the offer removal confirmation flow that prunes orphaned fragments from selection.

File: studio/src/promotions/promotion-editor-utils.js (hunk 3 above)
Comment (Andrei4226) [reply studio/src/promotions/promotion-editor-utils.js:308]: Both functions are used in `mas-promotions-items-table.js`.

File: studio/src/promotions/promotion-editor-utils.js (hunk 4 above)
Comment (Andrei4226) [reply studio/src/promotions/promotion-editor-utils.js]: Checked.
I think we don't have an equivalent found in card editor.
Files that call `resolveOfferSelectors` elsewhere (e.g. `aem-tag-picker-field.js`) only extract `productArrangementCode` for a different purpose. 
Card editor reads the stored `mnemonicIcon` field directly and no derives icon URLs from an OST offer payload. Proactively extracted into `offer-utils.js` anyway for separation of concerns.

File: studio/src/promotions/promotion-editor-utils.js (hunk 6 above)
Comment (Andrei4226) [reply studio/src/promotions/promotion-editor-utils.js]: Done.
I renamed all `Cache/Snapshot` words to `Record/Fields` throughout `promotion-editor-utils.js`, `mas-promo-codes-manager.js`, `mas-promotions-editor.js` and all related test files.

File: studio/src/promotions/mas-promotions-editor.js (hunk 7 above)
Comment (Andrei4226) [reply studio/src/promotions/mas-promotions-editor.js:34]: Some of these functions are used outside `mas-promotions-editor`.
for e.g. `splitPromotionTagsFieldValues` is imported in `mas-fragment-editor.js`, `isPromotionItemSelectionDirty/isPromotionOffersSelectionDirty` in `router.js` and `PROMOTION_FIELD_TYPE_MAP` in `mas-promotions.js`.

File: studio/src/promotions/mas-promotions.js (hunk 8 above)
Comment (Andrei4226) [reply studio/src/promotions/mas-promotions.js:19]: Both `mas-promotions.js` and `mas-promotions-editor.js` import from `promotion-publish-utils`, so moving it into `mas-promotions.js` would require `mas-promotions-editor.js` to import from `mas-promotions.js`.
I think this change would create an awkward reverse dependency but we can discuss

**Comment by Andrei4226** [comment]
> 2 main requested changes, rest is more "cosmetic":
> 
> * osi substitutions should be made also within the content of the fragment. Fragment's osi is here for "categorizing" it. It's later used in card edition to set prices & all. Those prices need to be updated.
> * we'd need to double check if it's MVP, but once i added an offer to my project, i should be able to add another one.
> 
> image i understand that this kind of make the fragment search complicated (maybe a dropdown applying a tag set or another depending on which offer are you search fragments for)
> but yeah, you could have e.g. promo project for CCI & Photoshop.

Thanks for your comments! Both done:
 1. OSI substitution now applied in wcs.js at WCS fetch time so substituteOsi() covers both data-wcs-osi attributes in fragment HTML content and body.fields.osi
 2. Multi-offer support added. A filter dropdown appears in the fragment picker when more and equal with 2 offers are selected, defaulting to "All offers"(selected offers).
 
image

File: io/www/src/fragment/transformers/wcs.js
Comment (npeltier) [inline io/www/src/fragment/transformers/wcs.js:120]: @Andrei4226 we are on the good path! but this is still not enough:

here you will fill the cache with the right substituted osi, but body will still be the same, so body will have old osi, mas/io payload will have new osi in wcs subobject. What will happen on client is that old osi will be used, and cache won't have the old osi so just fetch the response to wcs. 

you need to rewrite in that process the full body, with new osi. match object will give you the exact places where to cut before & after the osi. So you can generate on the fly a new bodyString and at the end of that transformer replace it with it.

Comment (Andrei4226) [reply io/www/src/fragment/transformers/wcs.js:120]: Done and hope I understood correctly now.
I iterated the matches in order, tracked a cumulative offset as each replacement shifts the string and used `match.index + offset` to cut before/after each OSI
`context.body = JSON.parse(newBodyString)` replaces the body

File: io/www/test/fragment/wcs.test.js [hunk 9]
```diff
@@ -241,3 +241,132 @@ describe('wcs corner cases', function () {
...
+        context.substituteMap = { 'BASE-OSI': 'SUB-OSI' };
+        delete context.wcsConfiguration;
+
+        context = await wcs.process(context);
+
+        expect(context.body.prices).to.include('data-wcs-osi="SUB-OSI"');
+        expect(context.body.prices).to.not.include('data-wcs-osi="BASE-OSI"');
+        expect(context.body.wcs).to.be.undefined;
+    });
+});
```
Comment (npeltier) [inline io/www/test/fragment/wcs.test.js:405]: please add a unit test with more than one data-wcs-osi placeholder, looking at your code, i suspect there is an issue

**Comment by npeltier** [comment]
when managing offers, we should not have promo code fields fed with default project promo code. It should be empty, and fed only if you want for that geo/offer, to override the promo code
image

File: io/www/src/fragment/transformers/wcs.js [hunk 10]
```diff
@@ -61,20 +65,45 @@ async function computeCache(tokens, wcsContext) {
...
+            const matchStart = match.index + offset;
+            const origMatchStr = match[0];
+            const newMatchStr = origMatchStr.replace(`\\"${originalOsi}\\"`, `\\"${substitutedOsi}\\"`);
+            newBodyString =
+                newBodyString.slice(0, matchStart) + newMatchStr + newBodyString.slice(matchStart + origMatchStr.length);
+            offset += newMatchStr.length - origMatchStr.length;
+        }
+    });
+    if (newBodyString !== bodyString) {
+        context.body = JSON.parse(newBodyString);
```
Comment (npeltier) [inline io/www/src/fragment/transformers/wcs.js]: please catch parse error, cancelling any change with an error log in case this happens

File: io/www/src/fragment/transformers/wcs.js [hunk 11]
```diff
@@ -61,20 +65,45 @@ async function computeCache(tokens, wcsContext) {
...
+    let offset = 0;
+    matches.forEach((match) => {
+        const originalOsi = match.groups.osi;
+        const substitutedOsi = substituteOsi(originalOsi, context.substituteMap);
+        if (originalOsi !== substitutedOsi) {
+            const matchStart = match.index + offset;
+            const origMatchStr = match[0];
+            const newMatchStr = origMatchStr.replace(`\\"${originalOsi}\\"`, `\\"${substitutedOsi}\\"`);
+            newBodyString =
+                newBodyString.slice(0, matchStart) + newMatchStr + newBodyString.slice(matchStart + origMatchStr.length);
```
Comment (npeltier) [inline io/www/src/fragment/transformers/wcs.js]: i wonder if concatenating ongoing transformed body with string from lastMatchEnd to matchStart + newMatchStr everytime is not more efficient?

**Review by npeltier — CHANGES_REQUESTED** [review]

File: io/www/src/fragment/transformers/wcs.js [hunk 12]
```diff
@@ -61,20 +65,47 @@ async function computeCache(tokens, wcsContext) {
...
+        } catch (error) {
+            logError(`Failed to rewrite body with substituted OSIs: ${error.message}`, context);
+        }
+    }
+    if (context.body.fields?.osi) {
+        const substitutedOsi = substituteOsi(context.body.fields.osi, context.substituteMap);
+        if (substitutedOsi !== context.body.fields.osi) {
+            context.body.fields.osi = substitutedOsi;
+        }
+    }
```
Comment (npeltier) [inline io/www/src/fragment/transformers/wcs.js:93]: ho i see you moved osi substitution here, fine by me i think

**Review by npeltier — APPROVED** [review]
approving as a developer :-) will wear my QA hat now

File: io/www/src/fragment/transformers/promotions.js [hunk 13]
```diff
@@ -119,15 +118,65 @@ function toInstant(value) {
...
+        if (!sub.geo) {
+            map[sub.baseOsi] = sub.substituteOsi;
+            continue;
+        }
+        const cqGeo =
+            sub.geo.startsWith('mas:') || sub.geo.startsWith('/content/')
+                ? sub.geo
+                : sub.geo.includes('_')
+                  ? `mas:locale/${sub.geo}`
+                  : `mas:country/${sub.geo}`;
```
Comment (npeltier) [inline io/www/src/fragment/transformers/promotions.js]: geo should only be having tag form. Please remove that part

**Review by 3ch023 — APPROVED** [review]
